### PR TITLE
[AppGate] Implement DelegateToGateway and add Tests

### PR DESCRIPTION
**Pull request by h5law**
### Human Summary

- Implement DelegateToGateway message and message handlers
- Add `DelegateePubKeys` field to the application proto
- Add extensive tests

### Summary generated by Reviewpad on 31 Oct 23 20:04 UTC
This pull request includes several changes across multiple files. 

Here's a summary of the changes:

1. `gateway.go`:
   - Changes related to the `GatewayKeeper` function.
   - Removed a line that initializes a variable.

2. `go.mod`:
   - Added dependencies on `github.com/cosmos/cosmos-proto` and `google.golang.org/genproto`.
   - Removed dependencies on `github.com/cosmos/cosmos-proto` and `google.golang.org/genproto`.

3. `delegate_to_gateway.go`:
   - Modified the `SimulateMsgDelegateToGateway` function.
   - Added imports for "pocket/x/application/keeper" and "pocket/x/application/types".
   - Added two new variables `simAppAccount` and `simGatewayAccount`.
   - Modified the `msg` initialization by replacing `Address` with `AppAddress` and `GatewayAddress`.
   - Added a TODO comment indicating that the handling of the `DelegateToGateway` simulation is yet to be implemented.

4. `delegate_to_gateway_test.go`:
   - Test functions for the `DelegateToGateway` method in the `MsgServer` implementation of the `keeper` package.
   - Tests for various scenarios such as successful delegation, duplicate delegation, and delegation to an unstaked gateway.

5. `tx.proto`:
   - Changes related to the `MsgDelegateToGateway` message.
   - Added an option and two fields related to address encoding and gateway delegation.

6. `stake_application.go`:
   - Changes related to validation for stake values and gateway addresses.
   - Ensures the validity of the stake value and gateway addresses for the applications.

7. `trust_to_gateway.sh`:
   - Adds new targets for delegating trust to different gateways.
   - Targets are named with the format "<app-name>_delegate_<gateway-name>".

8. `gateway_test.go`:
   - Adds import statement for "pocket/x/gateway/types".
   - Adds new variables and functions related to mocking and testing.
   - Updates the initialization of the `Keeper` struct with mock implementations.

9. `expected_keepers.go`:
   - Changes in the `types` package.
   - Updates the mock generation command for the `GatewayKeeper` interface.
   - Adds import statement for the `gatewaytypes` package.
   - Defines and imports the `GatewayKeeper` interface with a method `GetGateway`.

10. `errors.go`:
    - Changes related to error handling.
    - Adds new error variables for different scenarios.

11. `app.go`:
    - Changes related to the application client CLI.
    - Updates import statements.
    - Modifies the `CmdDelegateToGateway` function to provide usage information.
    - Changes the `Args` field of the `CmdDelegateToGateway` command to require one argument.
    - Modifies the `RunE` function of the `CmdDelegateToGateway` command to retrieve the gateway address.
    - Adds validation for the `gatewayAddress` parameter.

12. `keeper.go`:
    - Changes related to the `keeper` package.
    - Adds new parameters `accountKeeper` and `gatewayKeeper` to the `NewKeeper` function.

13. `application.proto`:
    - Adds a new field for delegatee gateway addresses in the `Application` message.

14. `types.go`:
    - Changes related to the `keeper` package in the `x/application` directory.
    - Adds `accountKeeper` and `gatewayKeeper` variables.
    - Updates the `NewKeeper` function to include the new parameters.

15. `message_delegate_to_gateway_test.go`:
    - Changes in the `message_delegate_to_gateway_test.go` file.
    - Adds test cases for different scenarios related to app and gateway addresses.

16. `message_delegate_to_gateway.go`:
    - Changes related to address parameters and validation.

17. `keeper.go` (in `x/gateway/keeper`):
    - Changes related to the `keeper` package in the `x/gateway/keeper` directory.
    - Removes the `accountKeeper` dependency from the `types` package.
    - Removes the `accountKeeper` parameter from the `NewKeeper` function.

18. `genesis_test.go`:
    - Changes related to the `genesis_test.go` file.
    - Adds new variables.
    - Modifies existing test cases and adds new test cases.

Please review these changes and let me know if you have any further questions or if there's anything specific you would like me to focus on.

## Issue

Fixes in part [39](https://github.com/pokt-network/poktroll/issues/39) 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_test`
- [x] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: x/application/keeper/msg_server_delegate_to_gateway.go
Comment (h5law) [inline x/application/keeper/msg_server_delegate_to_gateway.go:20]: @Olshansk WDYT here? I have it set up that the app must already be staked for it to delegate and I am confident this is the right way to do things. LMK if you think differently

File: proto/pocket/application/application.proto
Comment (Olshansk) [inline proto/pocket/application/application.proto:17]: Want to raise a few points:

1. Allow this to be nullable (i.e. you don't have to delegate)
2. Use a repeated string of addresses instead of repeated pubkeys (shorter & simpler) + you have access to the account keeper
3. s/delegateePubKeys/gatewayPubKeys or s/delegateePubKeys/gatewayAddresses so its easier to read and update the comment to explained its delegated to gateways

File: proto/pocket/application/tx.proto
Comment (Olshansk) [inline proto/pocket/application/tx.proto:38]: 🐍 💼 
Screenshot 2023-10-26 at 4 01 28 PM

File: testutil/keeper/application.go
Comment (Olshansk) [inline testutil/keeper/application.go:29]: Is this necessary? I think the declaration on its own achieves the same thing.

File: testutil/keeper/application.go
Comment (Olshansk) [inline testutil/keeper/application.go:52]: Nice comment

File: testutil/keeper/application.go
Comment (Olshansk) [inline testutil/keeper/application.go:54]: Move `func(...)` to the next line and indent things appropriately for redability.

Ditto below.

File: x/application/client/cli/tx_delegate_to_gateway.go
Comment (Olshansk) [inline x/application/client/cli/tx_delegate_to_gateway.go:28]:
```suggestion

			gatewayAddress := args[0]
```

File: x/application/types/genesis.go
Comment (Olshansk) [inline x/application/types/genesis.go:53]: s/delegatee/gatewayPubKey if we go with the rename

File: x/application/keeper/msg_server_stake_application.go
Comment (Olshansk) [inline x/application/keeper/msg_server_stake_application.go:67]: Not a fan of the fact that we have to use `Any` in such a core piece of the protocol.

File: x/application/client/cli/tx_delegate_to_gateway_test.go
Comment (Olshansk) [inline x/application/client/cli/tx_delegate_to_gateway_test.go:98]: outDel looks like delete :)

File: x/application/types/utils.go
Comment (Olshansk) [inline x/application/types/utils.go:16]: Can these 3 lines be part of an init function?

File: x/application/keeper/msg_server_delegate_to_gateway.go
Comment (Olshansk) [inline x/application/keeper/msg_server_delegate_to_gateway.go:16]: Make sure to revalidate Basic the msg. 

This is a state transition.

File: x/application/keeper/msg_server_delegate_to_gateway.go
Comment (Olshansk) [reply x/application/keeper/msg_server_delegate_to_gateway.go]: **tl;dr Yes.**

Fundamentally, it doesn't really matter because. If an app delegates without staking, it can't use relays anyhow. However, I just think things will be simpler to test/verify if it's a requirement so let's go with what you have.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Overall, REALLY great work.

Aside from the comments. There are three major things I want to discuss:

1. Consider rename `delgateeX` to `gatewayX`. Wdyt?

2. Why we're storing `pubKeys` instead of `address` in the `app` object. Why did you take this approach?

3. Push back against using `Any` instead of a well-structured type. Can you give more details on the reasoning/issue here?

File: proto/pocket/application/application.proto
Comment (h5law) [reply proto/pocket/application/application.proto:17]: > Allow this to be nullable (i.e. you don't have to delegate)

Nullable causes serialisation errors. The null form of this is an empty array. We want this always present so that the app can sign its own relays off chain and the servicers can build the ring easily (ring only with app pub key). Introducing nullable slice here caused issues in the alpha repo.

> Use a repeated string of addresses instead of repeated pubkeys (shorter & simpler) + you have access to the account keeper

I can do this, as long as the off chain actors can rebuild the ring from the addresses (also has access to the account keeper) this shouldnt be a problem

> s/delegateePubKeys/gatewayPubKeys or s/delegateePubKeys/gatewayAddresses so its easier to read and update the comment to explained its delegated to gateways

makes sense 👍🏼

File: proto/pocket/application/tx.proto
Comment (h5law) [reply proto/pocket/application/tx.proto:38]: We have gone over this before. Cosmos standard is to use the camelCase and their compiler supports it. Should I change these?

File: testutil/keeper/application.go
Comment (h5law) [reply testutil/keeper/application.go:29]: This is necessary go doesn't like an uninitiated map iirc

File: proto/pocket/application/application.proto
Comment (h5law) [reply proto/pocket/application/application.proto:17]: @Olshansk on second thought, as we spoke yesterday offline about not auto-undelegating when a Gateway unstakes, this would make the retrieval of the gateway's public key from their address potentially impossible (if it doesn't exist on chain we cant find the pubkey) and thus the ring cannot be built.

The `Any` type used by cosmos is a customised version of the Any protobuf message that is well supported and this is the standard way of encoding public keys in a protobuf. I would push to keep public keys until this can be worked around

Comment (h5law) [reply proto/pocket/application/application.proto:17]: Also I went with `DelegateeGatewayPubKey` to be verbose as hell

File: x/application/keeper/msg_server_stake_application.go
Comment (h5law) [reply x/application/keeper/msg_server_stake_application.go:67]: `Any` is the cosmos standard way of doing pubkey encoding. Their `Any` type is a modified version of the protobuf message. Cosmos uses it internally for this exact feature which is why I am using it

File: x/application/types/utils.go
Comment (bryanchriswhite) [inline x/application/types/utils.go:1]: I think we should move and rename this file in to an appropriate go pkg. Naming things "utils" is typically a code smell IMHO. Wdyt about something like `poktroll/pkg/crypto/codec.go`?

File: x/application/types/utils.go
Comment (bryanchriswhite) [inline x/application/types/utils.go:30]: 1. Perhaps this should be in its own file (e.g. `key.go`)
2. 
    ```suggestion
    // PublicKeyToBech32 converts a cryptotypes.PubKey to a bech32 address string
    func PublicKeyToBech32(publicKey cryptotypes.PubKey) string {
    ```

File: x/application/types/utils.go
Comment (h5law) [reply x/application/types/utils.go:1]: I would review #112 and probably merge it into this repo as it remove this entirely among other changes that seemed worth its own PR

File: proto/pocket/application/application.proto
Comment (bryanchriswhite) [inline proto/pocket/application/application.proto:17]: 1. I was following this comment until the "for use ..." part. To me, it sounds like it's referring to some other place in code but it also sounds like it's describing this field, which I find a bit confusing; to say something like "this field holds X for use in this field" or "...where this field is used". Perhaps I'm misunderstanding but does that make sense?
2. #PUC on what the `[(cosmos_proto.accepts_interface)]` option is and/or where to learn more.

File: proto/pocket/application/tx.proto
Comment (bryanchriswhite) [reply proto/pocket/application/tx.proto:38]: We have been preferring to conform with protobuf convention over cosmos on this point and I think we should continue to do so, if for no other reason than consistency.

Comment (bryanchriswhite) [reply proto/pocket/application/tx.proto:38]:
```suggestion
  string appAddress = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"]; // The Bech32 address of the application using cosmos' ScalarDescriptor to ensure deterministic encoding.
```

File: proto/pocket/application/tx.proto
Comment (bryanchriswhite) [inline proto/pocket/application/tx.proto:39]:
```suggestion
  string gatewayAddress = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"]; // The Bech32 address of the gateway the application wants to delegate to; using cosmos' ScalarDescriptor to ensure deterministic encoding .
```

File: proto/pocket/application/tx.proto
Comment (bryanchriswhite) [reply proto/pocket/application/tx.proto:38]: Wdyt about moving comments in protobufs to be above the lines they comment on, as we do in go source? (cc @Olshansk)

I think it would improve readability, especially for long lines like these.

File: testutil/keeper/application.go
Comment (bryanchriswhite) [inline testutil/keeper/application.go:30]: Can this not be simplified to?:

```suggestion
var AddrToPubKeyMap = make(map[string]cryptotypes.PubKey)
```

File: x/application/client/cli/tx_delegate_to_gateway.go
Comment (bryanchriswhite) [inline x/application/client/cli/tx_delegate_to_gateway.go:22]: I :heart: to see the long description! :rocket: 

I struggled with the word "over" in the phrase "delegate trust over". With some help from chatGPT I would propose rephrasing like so:

```suggestion
		Long: `Delegate an application to the gateway with the provided address. This is a broadcast operation
that delegates authority to the gateway specified to sign relays requests for the application, allowing the gateway
act on the behalf of the application during a session.
```

File: x/application/client/cli/tx_delegate_to_gateway.go
Comment (bryanchriswhite) [inline x/application/client/cli/tx_delegate_to_gateway.go:18]: Not sure if this is a NIT or not but I wonder if it's worth specifying that this is a bech32 address?

```suggestion
		Use:   "delegate-to-gateway [gateway bech32 address]",
```

File: proto/pocket/application/application.proto
Comment (Olshansk) [reply proto/pocket/application/application.proto:9]: Do you have a linter that auto formats it?

File: testutil/keeper/application.go
Comment (Olshansk) [reply testutil/keeper/application.go:29]: I tested and confirmed that this is correct.

File: proto/pocket/application/application.proto
Comment (Olshansk) [reply proto/pocket/application/application.proto]: Resolving this (and will start a new thread if needed) per the numerous offline discussions.

**Review by Olshansk — COMMENTED** [review]
Looks pretty good to me. 

@h5law Before I approve, wanted to check in if you were planning to s/pubKey/address in this PR or the next one?

File: proto/pocket/application/application.proto
Comment (bryanchriswhite) [reply proto/pocket/application/application.proto:9]: I do not. I do it manually and check any time I'm staging changes in git. There is such a thing though. Looks [like it's part of `golanglint-ci`](https://golangci-lint.run/usage/linters/#goimports) but I don't think that applies to protobufs.

File: proto/pocket/application/tx.proto
Comment (Olshansk) [reply proto/pocket/application/tx.proto:38]: Reopening this thread to change these fields, and created #119 for future work.

---

@h5law  `Question`: Can you point to where this is enforced in their proto compiler?

> We have gone over this before. Cosmos standard is to use the camelCase and their compiler supports it. Should I change these?

I followed this thread in Cosmos (https://github.com/cosmos/cosmos-sdk/issues/6826) and it looks like they disabled automatic CamelCasing when interacting with JSON (https://github.com/cosmos/cosmos-sdk/pull/6833). I realize this is Cosmos' standard, but I think we should be consistent within our own codebase and protobuf general standards.

---

@bryanchriswhite Not going to block the comment structure here but added it as a listing check in #119.

File: x/application/types/message_delegate_to_gateway_test.go
Comment (Olshansk) [inline x/application/types/message_delegate_to_gateway_test.go:21]:
```suggestion
				// GatewayAddress: intentionally omitted
```

File: x/application/types/message_delegate_to_gateway_test.go
Comment (Olshansk) [inline x/application/types/message_delegate_to_gateway_test.go:28]:
```suggestion
				// GatewayAddress: intentionally omitted
```

File: x/application/client/cli/tx_delegate_to_gateway.go
Comment (Olshansk) [reply x/application/client/cli/tx_delegate_to_gateway.go:11]: @h5law bump

File: x/application/client/cli/tx_delegate_to_gateway.go
Comment (Olshansk) [reply x/application/client/cli/tx_delegate_to_gateway.go:18]: I think it's okay to avoid the `bech32` mention here. Might end up being too verbose

File: x/application/client/cli/tx_delegate_to_gateway_test.go
Comment (Olshansk) [inline x/application/client/cli/tx_delegate_to_gateway_test.go:77]:
```suggestion
			desc:           "delegate to gateway: valid",
			appAddress:     appAccount.Address.String(),
			gatewayAddress: gatewayAccount.Address.String(),
		},
		{
			desc: "invalid - missing app address",
			// appAddress:     appAccount.Address.String(),
			gatewayAddress: gatewayAccount.Address.String(),
			err:            types.ErrAppInvalidAddress,
		},
		{
			desc:           "invalid - invalid app address",
			appAddress:     "invalid address",
			gatewayAddress: gatewayAccount.Address.String(),
			err:            types.ErrAppInvalidAddress,
		},
		{
			desc:       "invalid - missing gateway address",
			appAddress: appAccount.Address.String(),
			// gatewayAddress: gatewayAccount.Address.String(),
			err: types.ErrAppInvalidGatewayAddress,
		},
		{
			desc:           "invalid - invalid gateway address",
			appAddress:     appAccount.Address.String(),
			gatewayAddress: "invalid address",
			err:            types.ErrAppInvalidGatewayAddress,
		},
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Review by Olshansk — APPROVED** [review]

**Pull request merged**
